hera_cal/abscal.py

```diff
@@ -532,9 +602,63 @@ def ant_dly_phi_gain_arr(self):
         else:
             return None
 
+    @property
+    def dly_slope(self):
+        """ return _dly_slope dict """
+        if hasattr(self, '_dly_slope'):
+            return copy.deepcopy(self._dly_slope)
+        else:
+            return None
+
+    @property
+    def dly_slope_gain(self):
+        """ form complex gain from _dly_slope dict """
+        if hasattr(self, '_dly_slope'):
+            dly_slope = self.dly_slope
+            return odict(map(lambda k: (k, np.exp(2j*np.pi*self.freqs.reshape(1, -1)*np.einsum("i...,i->...", dly_slope[k], self.antpos[k[0]][:2]))), flatten(self._gain_keys)))
+        else:
+            return None
+
+    def custom_dly_slope_gain(self, gain_keys, antpos):
+        """
+        write dly_slope_gain with custom gain keys
+
+        gain_keys : type=list, list of unique (ant, pol). Ex. [(0, 'x'), (1, 'x'), (0, 'y'), (1, 'y')]
+        antpos : type=dictionary, contains antenna position vectors. keys are ant integer, values are ant positions
+        """
+        if hasattr(self, '_dly_slope'):
+            dly_slope = self.dly_slope[self._gain_keys[0][0]]
+            return odict(map(lambda k: (k, np.exp(2j*np.pi*self.freqs.reshape(1, -1)*np.einsum("i...,i->...", dly_slope, antpos[k[0]][:2]))), gain_keys))
+        else:
+            return None
+
+    @property
+    def dly_slope_arr(self):
+        """ return _dly_slope_arr array """
+        if hasattr(self, '_dly_slope_arr'):
+            return copy.copy(self._dly_slope_arr)
+        else:
+            return None
+
+    @property
+    def dly_slope_gain_arr(self):
+        """ form complex gain from _dly_slope_arr array """
+        if hasattr(self, '_dly_slope_arr'):
+            return np.exp(2j*np.pi*self.freqs.reshape(-1, 1)*np.einsum("hi...,hi->h...", self._dly_slope_arr, self.antpos_arr[:, :2]))
```

Comment:
This is line is too long. Also, can you add some comments to explain what's going on here...it's really hard to read.

Reply:
I agree these lines are long and a little cumbersome. It isn't very necessary for them to be "readable" per-se because they are not really exposed to the user, but better documentation will certainly help readability. The reason why I have it compressed to a single line is so that I don't have to perform unnecessary FOR loops.

Reply:
An acceptable solution is to just add a linebreak (you don't have to use for loops) and to add a comment explaining exactly what's going on. You could also separate the exponentiation from the einsum.